### PR TITLE
[#168821506] Profile email verification process

### PR DESCRIPTION
**Pull request by francescopersico**

### CODE REVIEW

File: openapi/index.yaml [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+  PaginationRequest:
+    name: cursor
+    in: query
+    type: string
+    minimum: 1
+    description: An opaque identifier that points to the next item in the collection.
+    x-example: 01BX9NSMKVXXS5PSP2FATZM123
+definitions:
+  ExtendedProfile:
+    $ref: "https://raw.githubusercontent.com/teamdigitale/io-functions-commons/168821506-profile-email-validation/openapi/definitions.yaml#/ExtendedProfile"
```
Comment (francescopersico) [inline openapi/index.yaml]: Temp reference to a specific branch until the PR of io-functions-commons will be merged.

File: UpdateProfile/handler.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  return async (context, fiscalCode, profilePayload) => {
+    const logPrefix = `UpdateProfileHandler|FISCAL_CODE=${fiscalCode}`;
+
+    const errorOrMaybeExistingProfile = await profileModel.findOneProfileByFiscalCode(
+      fiscalCode
+    );
+
+    if (isLeft(errorOrMaybeExistingProfile)) {
+      return ResponseErrorQuery(
+        "Error trying to retrieve existing service",
```
Comment (gunzip) [inline UpdateProfile/handler.ts]:
```suggestion
        "Error trying to retrieve existing profile",
```

File: UpdateProfile/handler.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      );
+      return ResponseErrorQuery(
+        "Error while updating the existing profile",
+        errorOrMaybeUpdatedProfile.value
+      );
+    }
+
+    const maybeUpdatedProfile = errorOrMaybeUpdatedProfile.value;
+
+    if (maybeUpdatedProfile.isNone()) {
```
Comment (gunzip) [inline UpdateProfile/handler.ts]:
```suggestion
    if (isNone(maybeUpdatedProfile)) {
```

(I'd chose one between the class or functional style)

File: UpsertProfileOrchestrator/handler.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+  const upsertProfileOrchestratorInput =
+    errorOrUpsertProfileOrchestratorInput.value;
+
+  const { newProfile, oldProfile, updatedAt } = upsertProfileOrchestratorInput;
+
+  const profileOperation = oldProfile !== undefined ? "UPDATED" : "CREATED";
+
+  // Check if the profile email is changed
+  if (profileOperation === "UPDATED" && newProfile.email !== oldProfile.email) {
```
Comment (gunzip) [inline UpsertProfileOrchestrator/handler.ts]:
```suggestion
  const isProfileEmailChanged = profileOperation === "UPDATED" && newProfile.email !== oldProfile.email;
  if (isProfileEmailChanged) {
```

File: templates/html/default.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+// DO NOT EDIT THIS FILE
```
Comment (gunzip) [inline templates/html/default.ts]: is this a copy of another file in any other functions project ?

File: utils/email.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import { Either, left, right } from "fp-ts/lib/Either";
```
Comment (gunzip) [inline utils/email.ts]: I guess this should be in some commons package

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a description of your PR changes.
  </td>
    </tr>
  </tbody>
</table>



## Affected stories

  * 🌟 [#168821506](https://www.pivotaltracker.com/story/show/168821506): Quando nel profilo viene cambiata l'email deve partire il processo di verifica


New dependencies added: <a href='https://www.npmjs.com/package/html-to-text'><code>html-to-text</code></a>, <a href='https://www.npmjs.com/package/nodemailer'><code>nodemailer</code></a>, <a href='https://www.npmjs.com/package/@types/html-to-text'><code>&#64;types/html-to-text</code></a>, <a href='https://www.npmjs.com/package/@types/lolex'><code>&#64;types/lolex</code></a>, <a href='https://www.npmjs.com/package/@types/nodemailer'><code>&#64;types/nodemailer</code></a> and <a href='https://www.npmjs.com/package/lolex'><code>lolex</code></a>.


<h2><a href="https://github.com/werk85/node-html-to-text">html-to-text</a></h2>
<p>Author: Malte Legenhausen</p>
<p>Description: Advanced html to plain text converter</p>
<p>Homepage: <a href="https://github.com/werk85/node-html-to-text">https://github.com/werk85/node-html-to-text</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 7 years ago</td></tr><tr><td>Last Updated</td><td>30 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>36</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/he'><code>he</code></a>, <a href='https://www.npmjs.com/package/htmlparser2'><code>htmlparser2</code></a>, <a href='https://www.npmjs.com/package/minimist'><code>minimist</code></a> and <a href='https://www.npmjs.com/package/lodash'><code>lodash</code></a></td></tr><tr><td>Keywords</td><td>html, node, text, mail, plain and converter</td></tr>
</table>
This README is too long to show.



<h2><a href="https://nodemailer.com/">nodemailer</a></h2>
<p>Author: Andris Reinman</p>
<p>Description: Easy as cake e-mail sending from your Node.js applications</p>
<p>Homepage: <a href="https://nodemailer.com/">https://nodemailer.com/</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 8 years ago</td></tr><tr><td>Last Updated</td><td>12 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>220</td></tr><tr><td>Direct Dependencies</td><td></td></tr><tr><td>Keywords</td><td>Nodemailer</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Nodemailer

[Backers on Open Collective](#backers) [Sponsors on Open Collective](#sponsors)

[Nodemailer](https://nodemailer.com/about/)

Send e-mails from Node.js – easy as cake! 🍰✉️

[NPM](https://nodemailer.com/about/)

See [nodemailer.com](https://nodemailer.com/) for documentation and terms.

## Why version bump to 5?

Nodemailer changed from `dns.lookup()` to `dns.resolve` for resolving SMTP hostnames which might be backwards incompatible and thus the version bump. Nodemailer tries first `resolve4()` and if no match is found then `resolve6()` and finally reverts back to `lookup()`. Additionally found DNS results are cached (for 5 minutes). This should make it easier to manage high performance clients that send a lot of messages in parallel.

## Having an issue?

#### First review the docs

Documentation for Nodemailer can be found at [nodemailer.com](https://nodemailer.com/about/)

#### Nodemailer throws a SyntaxError for "..."

You are using older Node.js version than v6.0. Upgrade Node.js to get support for the spread operator

#### I'm having issues with Gmail

Gmail either works well or it does not work at all. It is probably easier to switch to an alternative service instead of fixing issues with Gmail. If Gmail does not work for you then don't use it.

#### I get ETIMEDOUT errors

Check your firewall settings. Timeout usually occurs when you try to open a connection to a port that is firewalled either on the server or on your machine

#### I get TLS errors

If you are running the code in your own machine, then check your antivirus settings. Antiviruses often mess around with email ports usage. Node.js might not recognize the MITM cert your antivirus is using.

#### I have a different problem

If you are having issues with Nodemailer, then the best way to find help would be [Stack Overflow](https://stackoverflow.com/search?q=nodemailer) or revisit the [docs](https://nodemailer.com/about/).

## Contributors

This project exists thanks to all the people who contribute. [[Contribute](CONTRIBUTING.md)].
<a href="graphs/contributors"></a>

## Backers

Thank you to all our backers! 🙏 [[Become a backer](https://opencollective.com/nodemailer#backer)]

<a href="https://opencollective.com/nodemailer#backers" target="_blank"></a>

## Sponsors

Support this project by becoming a sponsor. Your logo will show up on our README on Github with a link to your website. [[Become a sponsor](https://opencollective.com/nodemailer#sponsor)]

<a href="https://opencollective.com/nodemailer/sponsor/0/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/1/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/2/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/3/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/4/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/5/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/6/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/7/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/8/website" target="_blank"></a>
<a href="https://opencollective.com/nodemailer/sponsor/9/website" target="_blank"></a>

### License

Nodemailer is licensed under the **MIT license**

---

The Nodemailer logo was designed by [Sven Kristjansen](https://www.behance.net/kristjansen).


</details>




<h2><a href="http://npmjs.com/package/@types/html-to-text">&#64;types/html-to-text</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for html-to-text</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/html-to-text">http://npmjs.com/package/@types/html-to-text</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>about 2 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>14</td></tr><tr><td>Direct Dependencies</td><td></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/html-to-text`

# Summary
This package contains type definitions for html-to-text (https://github.com/werk85/node-html-to-text).

# Details
Files were exported from https://www.github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/html-to-text

Additional Details
 * Last updated: Mon, 21 Aug 2017 21:55:03 GMT
 * Dependencies: none
 * Global values: htmlToText

# Credits
These definitions were written by Eryk Warren <https://github.com/erykwarren>.


</details>




<h2><a href="http://npmjs.com/package/@types/lolex">&#64;types/lolex</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for lolex</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/lolex">http://npmjs.com/package/@types/lolex</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>about 2 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>21</td></tr><tr><td>Direct Dependencies</td><td></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/lolex`

# Summary
This package contains type definitions for lolex ( https://github.com/sinonjs/lolex ).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/lolex

Additional Details
 * Last updated: Tue, 19 Mar 2019 18:15:45 GMT
 * Dependencies: none
 * Global values: none

# Credits
These definitions were written by Wim Looman <https://github.com/Nemo157>, Josh Goldberg <https://github.com/joshuakgoldberg>, Rogier Schouten <https://github.com/rogierschouten>, Yishai Zehavi <https://github.com/zyishai>.


</details>




<h2><a href="http://npmjs.com/package/@types/nodemailer">&#64;types/nodemailer</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for Nodemailer</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/nodemailer">http://npmjs.com/package/@types/nodemailer</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>about 2 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>42</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/@types/node'><code>&#64;types/node</code></a></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/nodemailer`

# Summary
This package contains type definitions for Nodemailer (https://github.com/nodemailer/nodemailer).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/nodemailer

Additional Details
 * Last updated: Mon, 22 Jul 2019 19:52:31 GMT
 * Dependencies: @types/node
 * Global values: none

# Credits
These definitions were written by Rogier Schouten <https://github.com/rogierschouten>, Piotr Roszatycki <https://github.com/dex4er>, and Daniel Chao <https://github.com/bioball>.


</details>




<h2><a href="http://github.com/sinonjs/lolex">lolex</a></h2>
<p>Author: Christian Johansen</p>
<p>Description: Fake JavaScript timers</p>
<p>Homepage: <a href="http://github.com/sinonjs/lolex">http://github.com/sinonjs/lolex</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>almost 5 years ago</td></tr><tr><td>Last Updated</td><td>about 2 hours ago</td></tr><tr><td>License</td><td>BSD-3-Clause</td></tr><tr><td>Maintainers</td><td>4</td></tr><tr><td>Releases</td><td>42</td></tr><tr><td>Direct Dependencies</td><td></td></tr>
</table>
This README is too long to show.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Pull request closed without merging**